File: app/lib/core/rental_service.dart
import 'package:flutter/foundation.dart';

import 'api_client.dart';

// ═════════════════════════════════════════════════════════════
// ── RentalService (singleton ChangeNotifier) ──────────────────
// ═════════════════════════════════════════════════════════════

class RentalData {
  final int id;
  final int pieceId;
  final String? pieceTitle;
  final int renterId;
  final String? renterUsername;
  final int ownerId;
  final String? ownerUsername;
  final String status;
  final DateTime startDate;
  final DateTime endDate;
  final int dailyRateCents;
  final int totalCents;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  RentalData({
    required this.id,
    required this.pieceId,
    this.pieceTitle,
    required this.renterId,
    this.renterUsername,
    required this.ownerId,
    this.ownerUsername,
    required this.status,
    required this.startDate,
    required this.endDate,
    required this.dailyRateCents,
    required this.totalCents,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  String get totalFormatted => '€${(totalCents / 100).toStringAsFixed(2)}';
  int get days => endDate.difference(startDate).inDays;

  factory RentalData.fromJson(Map<String, dynamic> j) => RentalData(
        id: j['id'] as int,
        pieceId: j['piece_id'] as int,
        pieceTitle: j['piece_title'] as String?,
        renterId: j['renter_id'] as int,
        renterUsername: j['renter_username'] as String?,
        ownerId: j['owner_id'] as int,
        ownerUsername: j['owner_username'] as String?,
        status: j['status'] as String,
        startDate: DateTime.parse(j['start_date'] as String),
        endDate: DateTime.parse(j['end_date'] as String),
        dailyRateCents: j['daily_rate_cents'] as int,
        totalCents: j['total_cents'] as int,
        notes: j['notes'] as String?,
        createdAt: DateTime.parse(j['created_at'] as String),
        updatedAt: DateTime.parse(j['updated_at'] as String),
      );
}

class BlockedDateRange {
  final DateTime startDate;
  final DateTime endDate;
  final String status;

  BlockedDateRange({
    required this.startDate,
    required this.endDate,
    required this.status,
  });

  factory BlockedDateRange.fromJson(Map<String, dynamic> j) => BlockedDateRange(
        startDate: DateTime.parse(j['start_date'] as String),
        endDate: DateTime.parse(j['end_date'] as String),
        status: j['status'] as String,
      );
}

class RentalService extends ChangeNotifier {
  static final RentalService _instance = RentalService._();
  static RentalService get instance => _instance;
  RentalService._();

  List<RentalData> _rentals = [];
  List<RentalData> get rentals => _rentals;

  bool _loading = false;
  bool get loading => _loading;

  /// Request a rental.
  Future<RentalData?> requestRental({
    required int pieceId,
    required DateTime startDate,
    required DateTime endDate,
    String? notes,
  }) async {
    try {
      final body = <String, dynamic>{
        'piece_id': pieceId,
        'start_date': startDate.toIso8601String().split('T')[0],
        'end_date': endDate.toIso8601String().split('T')[0],
      };
      if (notes != null) body['notes'] = notes;

      final data = await ApiClient.instance.post('/rentals', body);
      return RentalData.fromJson(data as Map<String, dynamic>);
    } catch (_) {
      return null;
    }
  }

  /// Fetch rentals.
  Future<void> fetchRentals({
    String role = 'renter',
    String? status,
    int offset = 0,
    int limit = 20,
  }) async {
    _loading = true;
    notifyListeners();

    try {
      final params = <String, String>{
        'role': role,
        'offset': offset.toString(),
        'limit': limit.toString(),
      };
      if (status != null) params['status'] = status;

      final query = params.entries.map((e) => '${e.key}=${e.value}').join('&');
      final data = await ApiClient.instance.get('/rentals?$query');
      _rentals = (data as List)
          .map((j) => RentalData.fromJson(j as Map<String, dynamic>))
          .toList();
    } catch (_) {
      // keep existing
    }

    _loading = false;
    notifyListeners();
  }

  /// Get rental detail.
  Future<RentalData?> fetchRental(int rentalId) async {
    try {
      final data = await ApiClient.instance.get('/rentals/$rentalId');
      return RentalData.fromJson(data as Map<String, dynamic>);
    } catch (_) {
      return null;
    }
  }

  /// Owner respond to rental request.
  Future<RentalData?> respondToRental(int rentalId, {required bool accept}) async {
    try {
      final data = await ApiClient.instance
          .patch('/rentals/$rentalId/respond', {'accept': accept});
      return RentalData.fromJson(data as Map<String, dynamic>);
    } catch (_) {
      return null;
    }
  }

  /// Owner update rental status.
  Future<RentalData?> updateRentalStatus(
    int rentalId, {
    required String status,
  }) async {
    try {
      final data = await ApiClient.instance
          .patch('/rentals/$rentalId/status', {'status': status});
      return RentalData.fromJson(data as Map<String, dynamic>);
    } catch (_) {
      return null;
    }
  }

  /// Get blocked dates for a piece in a given month.
  Future<List<BlockedDateRange>> fetchCalendar(
    int pieceId, {
    required int year,
    required int month,
  }) async {
    try {
      final data = await ApiClient.instance
          .get('/rentals/calendar/$pieceId?year=$year&month=$month');
      final map = data as Map<String, dynamic>;
      return (map['blocked_dates'] as List)
          .map((j) => BlockedDateRange.fromJson(j as Map<String, dynamic>))
          .toList();
    } catch (_) {
      return [];
    }
  }
}

File: backend/app/services/rental_service.py
from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from ..models.user import User
from ..repositories.rental_repository import RentalRepository
from ..repositories.settings_repository import SettingsRepository
from ..schemas.rental import (
    BlockedDateRange,
    RentalCalendarOut,
    RentalCreate,
    RentalOut,
    RentalRespondIn,
    RentalStatusUpdate,
)
from .notification_helper import notify


class RentalService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.repo = RentalRepository(db)

    def _rental_to_out(self, rental) -> RentalOut:
        piece = self.repo.get_piece(rental.piece_id)
        renter = self.repo.get_user_by_id(rental.renter_id)
        owner = self.repo.get_user_by_id(rental.owner_id)
        return RentalOut(
            id=rental.id,
            piece_id=rental.piece_id,
            piece_title=piece.title if piece else None,
            renter_id=rental.renter_id,
            renter_username=renter.username if renter else None,
            owner_id=rental.owner_id,
            owner_username=owner.username if owner else None,
            status=rental.status,
            start_date=rental.start_date,
            end_date=rental.end_date,
            daily_rate_cents=rental.daily_rate_cents,
            total_cents=rental.total_cents,
            notes=rental.notes,
            created_at=rental.created_at,
            updated_at=rental.updated_at,
        )

    def create_rental(self, user: User, data: RentalCreate) -> RentalOut:
        piece = self.repo.get_piece(data.piece_id)
        if piece is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Piece not found.",
            )

        if not piece.rental:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="This piece is not available for rental.",
            )

        if piece.user_id == user.id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Cannot rent your own piece.",
            )

        if data.start_date >= data.end_date:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="End date must be after start date.",
            )

        # Check availability
        if self.repo.has_overlap(data.piece_id, data.start_date, data.end_date):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Piece is not available for the selected dates.",
            )

        # Calculate total
        days = (data.end_date - data.start_date).days
        daily_rate = piece.rental_daily_rate_cents or piece.price_cents
        total = daily_rate * days

        rental = self.repo.create(
            piece_id=data.piece_id,
            renter_id=user.id,
            owner_id=piece.user_id,
            start_date=data.start_date,
            end_date=data.end_date,
            daily_rate_cents=daily_rate,
            total_cents=total,
            notes=data.notes,
        )

        # Notify owner (gated by rental_requests setting)
        settings_repo = SettingsRepository(self.db)
        owner_settings = settings_repo.get_by_user(piece.user_id)
        if not owner_settings or owner_settings.rental_requests:
            notify(
                self.db,
                user_id=piece.user_id,
                type="rental",
                title="New rental request",
                body=f"{user.username} wants to rent '{piece.title}'.",
                reference_id=rental.id,
                reference_type="rental",
            )

        return self._rental_to_out(rental)

    def list_rentals(
        self,
        user: User,
        role: str = "renter",
        status_filter: str | None = None,
        offset: int = 0,
        limit: int = 20,
    ) -> list[RentalOut]:
        rentals = self.repo.list_rentals(user.id, role, status_filter, offset, limit)
        return [self._rental_to_out(r) for r in rentals]

    def get_rental(self, user: User, rental_id: int) -> RentalOut:
        rental = self.repo.get_by_id(rental_id)
        if rental is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Rental not found.",
            )

        if rental.renter_id != user.id and rental.owner_id != user.id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not authorized.",
            )

        return self._rental_to_out(rental)

    def respond_to_rental(
        self, user: User, rental_id: int, data: RentalRespondIn
    ) -> RentalOut:
        rental = self.repo.get_by_id(rental_id)
        if rental is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Rental not found.",
            )

        if rental.owner_id != user.id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Only the owner can respond to rental requests.",
            )

        if rental.status != "pending":
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Can only respond to pending rentals.",
            )

        new_status = "accepted" if data.accept else "declined"
        self.repo.update(rental, {"status": new_status})

        # Notify renter (gated by rental_status_changes setting)
        settings_repo = SettingsRepository(self.db)
        renter_settings = settings_repo.get_by_user(rental.renter_id)
        if not renter_settings or renter_settings.rental_status_changes:
            notify(
                self.db,
                user_id=rental.renter_id,
                type="rental",
                title=f"Rental {new_status}",
                body=f"Your rental request was {new_status}.",
                reference_id=rental.id,
                reference_type="rental",
            )

        return self._rental_to_out(rental)

    def update_rental_status(
        self, user: User, rental_id: int, data: RentalStatusUpdate
    ) -> RentalOut:
        rental = self.repo.get_by_id(rental_id)
        if rental is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Rental not found.",
            )

        if rental.owner_id != user.id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Only the owner can update rental status.",
            )

        valid_statuses = ["active", "returned", "cancelled"]
        if data.status not in valid_statuses:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid status. Must be one of: {valid_statuses}",
            )

        self.repo.update(rental, {"status": data.status})

        # Notify renter of status change (gated by rental_status_changes setting)
        settings_repo = SettingsRepository(self.db)
        renter_settings = settings_repo.get_by_user(rental.renter_id)
        if not renter_settings or renter_settings.rental_status_changes:
            notify(
                self.db,
                user_id=rental.renter_id,
                type="rental_status",
                title=f"Rental {data.status}",
                body=f"Your rental status has been updated to '{data.status}'.",
                reference_id=rental.id,
                reference_type="rental",
            )

        return self._rental_to_out(rental)

    def get_calendar(self, piece_id: int, year: int, month: int) -> RentalCalendarOut:
        blocked = self.repo.get_blocked_dates(piece_id, year, month)
        ranges = [
            BlockedDateRange(
                start_date=r.start_date,
                end_date=r.end_date,
                status=r.status,
            )
            for r in blocked
        ]
        return RentalCalendarOut(piece_id=piece_id, blocked_dates=ranges)
